lib/rust/profiler/macros/src/lib.rs

```diff
@@ -326,47 +318,217 @@ fn start_profiler(
+    let ret_ty = match &func.sig.output {
+        syn::ReturnType::Default => quote::quote! { () },
+        syn::ReturnType::Type(_, ty) => ty.into_token_stream(),
+    };
+    let body = quote::quote! {{
+        #before_async
+        async move {
+            #start_async
+            let result: #ret_ty = #block;
+            #end_async
+            result
+        }
+    }};
+    let output_ty = match &func.sig.output {
+        syn::ReturnType::Default => quote::quote! { () },
+        syn::ReturnType::Type(_, ty) => ty.to_token_stream(),
+    };
+    let output_lifetime = syn::Lifetime::new("'__profiler_out", proc_macro2::Span::call_site());
+    let lifetimes = explicitize_lifetimes(&mut func.sig);
+    let output = if lifetimes.is_empty() {
+        quote::quote! {
+            -> impl std::future::Future<Output=#output_ty>
+        }
+    } else {
+        // Bound all input lifetimes on the output lifetime.
+        for param in &mut func.sig.generics.params {
+            if let syn::GenericParam::Lifetime(def) = param {
+                def.bounds.push(output_lifetime.clone());
+            }
+        }
+        // Add a definition for the output lifetime.
+        let lifetime_def = syn::LifetimeDef::new(output_lifetime.clone());
+        func.sig.generics.params.insert(0, syn::GenericParam::Lifetime(lifetime_def));
+        // Apply the output lifetime to the output.
+        quote::quote! {
+            -> impl std::future::Future<Output=#output_ty> + #output_lifetime
+        }
+    };
+    func.sig.asyncness = None;
+    func.sig.output = syn::parse2(output).unwrap();
+    func.block = syn::parse2(body).unwrap();
+}
+
+/// Make all lifetimes in function signature explicit.
+///
+/// Returns the lifetimes used in the function signature.
+fn explicitize_lifetimes(sig: &mut syn::Signature) -> Vec<syn::Lifetime> {
+    // Go through the args; find:
+    // - anonymous lifetime: '_
+    // - implicit lifetimes: &foo
+    // - explicit lifetimes: &'a
+    // Make all input lifetimes explicit:
+    // - Use new lifetime explicitly in arg list.
+    // - Define new lifetime in generic params.
+    let mut input_transformer = ExplicitizeInputLifetimes::default();
+    for input in &mut sig.inputs {
+        input_transformer.visit_fn_arg_mut(input);
+    }
+    let ExplicitizeInputLifetimes { new_lifetimes, existing_lifetimes } = input_transformer;
+    let mut all_lifetimes = existing_lifetimes;
+    all_lifetimes.extend_from_slice(&new_lifetimes);
+    let new_lifetimes =
+        new_lifetimes.into_iter().map(|lt| syn::GenericParam::Lifetime(syn::LifetimeDef::new(lt)));
+    sig.generics.params.extend(new_lifetimes);
+    // There are two cases where output lifetimes may be elided:
+    // - There's exactly one lifetime in the inputs.
+    // - There's a receiver with a lifetime.
+    // If either case occurs, make any implicit output lifetimes explicit.
+    let default_lifetime = if all_lifetimes.len() == 1 {
```

Comment:
Thank you for providing comments with clear explanation in the code! It's amazing!